### PR TITLE
Enable optional metadata tracking in LNLS

### PR DESCRIPTION
**Pull request by kevinchern**
Add feature to track subsamples (QPU results), subproblems (BQMs sent to QPU), and samples (burndown)

### CODE REVIEW

**Comment by kevinchern** [comment]
@jackraymond 👀 👀

File: hybrid/reference/lattice_lnls.py [hunk 1]
```diff
@@ -118,11 +119,24 @@ def LatticeLNLS(topology,
...
+            updates[k1] = state.get(k1, []) + [state.get(k2)]
+        state = state.updated(**updates)
+        return state
+
     qpu_branch = (hybrid.decomposers.SublatticeDecomposer()
                   | hybrid.QPUSubproblemExternalEmbeddingSampler(
                       qpu_sampler=qpu_sampler,
                       sampling_params=qpu_params0,
                       num_reads=qpu_params0['num_reads']))
+    if track_data:
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply hybrid/reference/lattice_lnls.py:138]: how about `track_iteration_data`? It tracks subproblems (BQMs), subsamples (samples of the subproblems), and samples (on the full problem)

File: tests/test_reference_samplers.py [hunk 2]
```diff
@@ -69,22 +69,52 @@ def __init__(self, broken_nodes=None, topology_type=None, qpu_scale=4, **config)
...
+            qpu_params=dict(chain_strength=2), reject_small_problems=False,
+            track_data=False
+            )
+
+        sampleset = LatticeLNLSSampler().sample(
+            bqm=bqm, problem_dims=(2,2,2), num_reads=num_reads,
+            qpu_sampler=MockDWaveSamplerGeneralization(), topology='cubic',max_iter=max_iter,
+            qpu_params=dict(chain_strength=2), reject_small_problems=False,
+            track_data=True
+            )
```
[the remark this answers is not in the document]
Comment (kevinchern) [reply tests/test_reference_samplers.py:107]: No, but I did now 😛

File: hybrid/reference/lattice_lnls.py [hunk 3]
```diff
@@ -83,6 +84,10 @@ def LatticeLNLS(topology,
                    parallel with the QPU; the best result is accepted on
                    each iteration.
 
+        track_iteration_data (bool, optional, default=False):
+            Flag to indicate whether to track samples, subproblems, and subsamples (samples for the
```
Comment (jackraymond) [inline hybrid/reference/lattice_lnls.py:88]: Add that the data is stored in the info field of the returned sampleset?

File: hybrid/reference/lattice_lnls.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (randomir) [reply hybrid/reference/lattice_lnls.py]: That's too generic. How about `track_qpu_branch`?

[the remark this answers is not in the document]
Comment (kevinchern) [reply hybrid/reference/lattice_lnls.py]: that's accurate

File: hybrid/reference/lattice_lnls.py (hunk 3 above)
Comment (kevinchern) [reply hybrid/reference/lattice_lnls.py]: added and described the correspondence between indices and num_reads

**Comment by kevinchern** [comment]
@randomir @jackraymond should tracking be placed after `SplatComposer()` if the intention is to capture the sample of the overall problem after the subproblem has been solved and used to update the original state?

**Comment by randomir** [comment]
> @randomir @jackraymond should tracking be placed after `SplatComposer()` if the intention is to capture the sample of the overall problem after the subproblem has been solved and used to update the original state? i.e., instead of line 143 in lattice_lnls.py, should it go after the `SplatComposer`s in the conditionals beginning line 145?

Yes, if you want to track `samples` after being updated with `subsamples` from QPU.

File: tests/test_reference_samplers.py [hunk 4]
```diff
@@ -94,16 +94,20 @@ def test_track_data(self):
...
-            qpu_sampler=MockDWaveSamplerGeneralization(), topology='cubic',max_iter=max_iter,
+            qpu_sampler=MockDWaveSampler(), topology='cubic',max_iter=max_iter,
             qpu_params=dict(chain_strength=2), reject_small_problems=False,
-            track_data=False
+            track_qpu_branch=False
             )
 
+        self.assertEquals(False, 'tracked_samples' in sampleset.info)
+        self.assertEquals(False, 'tracked_subsamples' in sampleset.info)
+        self.assertEquals(False, 'tracked_subproblems' in sampleset.info)
```
Comment (randomir) [inline tests/test_reference_samplers.py:104]: You might want to use [`assertFalse`](https://apps.docusign.com/send/home) or even better [`assertNotIn`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertNotIn) for this. One good reason to use a more specific assertion, generally, is better/more specific error message.

**Pull request merged**